compose/foundation/foundation/src/desktopTest/kotlin/androidx/compose/foundation/OnClickTest.kt

```diff
@@ -0,0 +1,481 @@
+        button = PointerButton.Secondary,
+        pointerMatcher = PointerMatcher.mouse(PointerButton.Secondary)
+    )
+
+    @Test
+    fun tertiaryClicks() = testClick(
+        button = PointerButton.Tertiary,
+        pointerMatcher = PointerMatcher.mouse(PointerButton.Tertiary)
+    )
+
+    @Test
+    fun backClicks() = testClick(
+        button = PointerButton.Back,
+        pointerMatcher = PointerMatcher.mouse(PointerButton.Back)
+    )
+
+    @Test
+    fun forwardClicks() = testClick(
+        button = PointerButton.Forward,
+        pointerMatcher = PointerMatcher.mouse(PointerButton.Forward)
+    )
+
+    @Test
+    fun clickWithTouch() = ImageComposeScene(
+        width = 100,
+        height = 100,
+        density = Density(1f)
+    ).use { scene ->
+        var clicksCount = 0
+
+        scene.setContent {
+            Box(
+                modifier = Modifier
+                    .onClick(matcher = PointerMatcher.touch) {
+                        clicksCount++
+                    }.size(10.dp, 20.dp)
+            )
+        }
+
+        scene.sendPointerEvent(PointerEventType.Move, Offset(0f, 0f), type = PointerType.Touch)
+        scene.sendPointerEvent(PointerEventType.Press, Offset(0f, 0f), type = PointerType.Touch)
+        scene.sendPointerEvent(PointerEventType.Release, Offset(0f, 0f), type = PointerType.Touch)
+        assertThat(clicksCount).isEqualTo(1)
+
+        scene.sendPointerEvent(PointerEventType.Move, Offset(5f, 5f), type = PointerType.Touch)
+        scene.sendPointerEvent(PointerEventType.Press, Offset(5f, 5f), type = PointerType.Touch)
+        scene.sendPointerEvent(PointerEventType.Release, Offset(5f, 5f), type = PointerType.Touch)
+        assertThat(clicksCount).isEqualTo(2)
+    }
+
+    private fun testDoubleClick(
+        pointerMatcher: PointerMatcher,
+        button: PointerButton
+    ) = runBlocking {
+        val density = Density(1f)
+        val viewConfiguration = DefaultViewConfiguration(density)
+        ImageComposeScene(
+            width = 100,
+            height = 100,
+            density = density
+        ).use { scene ->
+            var clicksCount = 0
+            var doubleClickCount = 0
+
+            scene.setContent {
+                Box(
+                    modifier = Modifier
+                        .onClick(
+                            matcher = pointerMatcher,
+                            onDoubleClick = {
+                                doubleClickCount++
+                            }
+                        ) {
+                            clicksCount++
+                        }.size(10.dp, 20.dp)
+                )
+            }
+
+            scene.sendPointerEvent(PointerEventType.Move, Offset(0f, 0f))
+            scene.sendPointerEvent(PointerEventType.Press, Offset(0f, 0f), button = button)
+            scene.sendPointerEvent(PointerEventType.Release, Offset(0f, 0f), button = button)
+            delay(viewConfiguration.doubleTapTimeoutMillis * 2)
```

Comment:
`FlushCoroutineDispatcher` wraps passed into `ImageSceneDispatcher`. I.e. if it would be `UnconfinedTestDispatcher`, the functionality of `FlushCoroutineDispatcher` just won't be used, because unconfined dispatcher performs all callbacks immediately, and on flush there won't be any callbacks.

> and we don't control effectDispatcher from ImageComposeScene now.

It should be controlled, because `effectDispatcher` in fact will be `UnconfinedTestDispatcher`. Theoretically it should work. If you just set this dispatcher in `ImageComposeScene` constructor, will the test pass? If we it doesn't pass, then we have a bug in `ImageCompoScene`, but we can fix it later.

Reply:
When I simply pass `UnconfinedTestDispatcher ` into `ImageComposeScene ` and try to advanceTime using that dispatcher in the test, the test fails. 

It seems, it has something to do with a fact that FlashCoroutineDispatcher doesn't implement a Delay needed in `withTimeoutOrNull`, so DefaultDelay is used - `Dispatchers.Default`, which doesn't know about UnconfinedTestDispatcher.

Reply:
Oh, I understand. We should fix that someday, because it also used in the usual Compose unit tests with `createComposeRule`. I will look onto this, when I will be fixing  the other ImageComposeScene bugs.

For now, let's just keep these tests as they are now.